File: android/app/src/main/java/com/fuggel/Uway/socket/SocketIOClient.kt
package com.fuggel.Uway.socket

import android.util.Log
import com.fuggel.Uway.constants.Constants
import io.socket.client.IO
import io.socket.client.Socket
import org.json.JSONObject
import java.net.URISyntaxException

class SocketIOClient(
    private val url: String = Constants.SOCKET_URL,
    private val authToken: String,
    private val onMessageReceived: (String) -> Unit,
) {
    private val tag = "SocketIOClient"
    private var socket: Socket? = null

    fun connect(onConnected: (() -> Unit)? = null) {
        try {
            val opts = IO.Options().apply {
                auth = mapOf("token" to authToken)
            }

            socket = IO.socket(url, opts)
        } catch (e: URISyntaxException) {
            Log.e(tag, "Invalid URL: $url", e)
            return
        }

        socket?.apply {
            on(Socket.EVENT_CONNECT) {
                onConnected?.invoke()
            }
            on(Socket.EVENT_DISCONNECT) {
                Log.d(tag, "Disconnected")
            }
            on(Socket.EVENT_CONNECT_ERROR) { args ->
                Log.e(tag, "Connect error: ${args.getOrNull(0)}")
            }
            on(Constants.WARNING_MANAGER) { args ->
                val message = args.getOrNull(0)?.toString() ?: return@on
                Log.d(tag, "Message received: $message")
                onMessageReceived(message)
            }

            connect()
            Log.d(tag, "Socket connected to $url")
        }
    }

    fun sendLocation(
        eventType: String,
        lon: Double,
        lat: Double,
        heading: Double? = null,
        speed: Double? = null
    ) {
        val json = JSONObject().apply {
            put("eventType", eventType)
            put("lon", lon)
            put("lat", lat)
            put("heading", heading ?: JSONObject.NULL)
            put("speed", speed ?: JSONObject.NULL)
            put("eventWarningType", JSONObject.NULL)
        }

        socket?.emit(Constants.USER_LOCATION, json)

        Log.d(tag, "Location sent: $json")
    }

    fun disconnect() {
        socket?.disconnect()
        socket?.off()
        socket = null
        Log.d(tag, "Socket disconnected and cleaned up")
    }
}

File: ios/Socket/SocketIOClient.swift
//
//  SocketIOClient.swift
//  Uway
//
//  Created by Furkan Ceylan on 20.06.25.
//

import Foundation
import SocketIO

class SocketClient {
    private var manager: SocketManager!
    private var socket: SocketIO.SocketIOClient!
    private let authToken: String
    
    var onMessageReceived: ((String) -> Void)?
    
    init(authToken: String) {
        self.authToken = authToken
        
        let url = URL(string: AppConfig.uwayWsUrl)!
        manager = SocketManager(
            socketURL: url,
            config: [
                .log(true),
                .compress,
                .connectParams(["token": authToken])
            ]
        )
        socket = manager.defaultSocket
    }
    
    func connect() {
        socket.on(clientEvent: .connect) { _, _ in
            print("Socket connected")
        }
        
        socket.on("warning") { [weak self] data, _ in
            if let message = data.first as? String {
                self?.onMessageReceived?(message)
            } else if let dict = data.first as? [String: Any],
                      let jsonData = try? JSONSerialization.data(withJSONObject: dict),
                      let jsonString = String(data: jsonData, encoding: .utf8) {
                self?.onMessageReceived?(jsonString)
            }
        }
        
        socket.on(clientEvent: .disconnect) { _, _ in
            print("Socket disconnected")
        }
        
        socket.on(clientEvent: .error) { data, _ in
            print("Socket error: \(data)")
        }
        
        socket.connect()
    }
    
    func sendLocation(
        eventType: String,
        lon: Double,
        lat: Double,
        heading: Double,
        speed: Double
    ) {
        let userId = UUID().uuidString
        let json: [String: Any] = [
            "eventType": eventType,
            "lon": lon,
            "lat": lat,
            "heading": heading,
            "speed": speed,
            "userId": userId,
            "eventWarningType": NSNull()
        ]
        
        socket.emit("location", json)
    }
    
    func disconnect() {
        socket.disconnect()
    }
}
